### ISSUE TITLE
Problems adding arbitrary headers through Request Signing

### ISSUE THREAD

**Issue by richardkerr** [swagger-api/swagger-js#591]
I'm trying to add arbitrary headers to a request and have been looking at the 'Custom Request Signing' example in the documentation, but I can't see how to register the Signer so that it is including in the processing chain?  
The only option, to me, seems to be to modify SwaggerClient so that the clientAuthorizations are replaced by the new signer, but it feels very wrong.  Help?

I'm coming from swagger-ui, if that makes any difference.

Thanks!

**Comment by fehguy** [comment]
No  need to modify code, you can add the custom signing logic either before initializing swagger-ui:

```
window.swaggerUi = new SwaggerUi({
   url: url,
   dom_id: "swagger-ui-container",
   authorizations: {
     api_key : new SwaggerClient.ApiKeyAuthorization("api_key", "foo", "query")
  },
```

Or after swagger-ui has been initialized (note! This will not help if the initial call requires signing):

https://github.com/swagger-api/swagger-ui#custom-header-parameters---for-basic-auth-etc

**Comment by richardkerr** [comment]
This requires the "api_key" security attribute to be available for the documented operation.  I want to be able to apply an 'authorization' (which is not actually a security related operation) to the object regardless of the documentation.

**Comment by richardkerr** [comment]
Some more info, it seems like the behaviour of the 'authorizations' option has changed a bit between versions.  
In 2.1.6-M2, it behaves like so:
```javascript
  if (options.authorizations) {
    this.clientAuthorizations = options.authorizations;
  } else {
    this.clientAuthorizations = new auth.SwaggerAuthorizations();
  }
```
This allows me to override the default implementation of SwaggerAuthorizations to have the custom signing as is documented.

In 2.1.6 and later the SwaggerClient constructor & initialization does:
```javascript
this.clientAuthorizations = new auth.SwaggerAuthorizations();
...
this.clientAuthorizations.add(options.authorizations);
```
This explains why the security attribute must be present, it's using SwaggerAuthorizations which does all this checking.  But it also means that we can't add an 'authorization' that is not documented.

This change was made in commit a769bc211d849337c5dba56534b97157c6534eb6, possibly an unintended side effect?

**Comment by richardkerr** [comment]
In 2.1.6+ it also, due to the behaviour of SwaggerAuthorizations, is applied when there are _no_ 'security' definitions in the documented operations.

**Comment by fehguy** [comment]
fixed in #632
